src/Std/Time/Notation.lean

```diff
@@ -0,0 +1,267 @@
+    `(Std.Time.PlainDate.mk $(← parseComponent none none year) $(← parseComponent (some 1) (some 12) month) $(← parseComponent (some 1) (some 31) day) (by decide))
+  | syn => Macro.throwErrorAt syn "unsupported type"
+
+/--
+Category of times like `HH:mm:ss`.
+-/
+declare_syntax_cat time
+
+/--
+Date in `HH-mm-ss` format.
+-/
+syntax date_component noWs ":" noWs date_component noWs ":" noWs date_component : time
+
+/--
+Date in `HH-mm-ss,sssssssss` format.
+-/
+syntax date_component noWs ":" noWs date_component noWs ":" noWs date_component noWs "," noWs date_component : time
+
+private def parseTime : TSyntax `time -> MacroM (TSyntax `term)
+  | `(time| $hour:date_component:$minute:date_component:$second:date_component,$nanos:date_component) => do
+    `(Std.Time.PlainTime.mk ⟨true, $(← parseComponent (some 0) (some 24) hour)⟩ $(← parseComponent (some 0) (some 59) minute) ⟨true, $(← parseComponent (some 0) (some 60) second)⟩ $(← parseComponent (some 0) (some 999999999) nanos) (by decide))
+  | `(time| $hour:date_component:$minute:date_component:$second:date_component) => do
+    `(Std.Time.PlainTime.mk ⟨true, $(← parseComponent (some 0) (some 24) hour)⟩ $(← parseComponent (some 0) (some 59) minute) ⟨true, $(← parseComponent (some 0) (some 60) second)⟩ 0 (by decide))
+  | syn => Macro.throwErrorAt syn "unsupported syntax"
+
+/--
+Category of date and time together.
+-/
+declare_syntax_cat datetime
+
+/--
+Date and time in `YYYY-MM-DDTHH:mm:ss,nano` format.
+-/
+syntax date &"T" time : datetime
+
+/--
+Date but using timestamp.
+-/
+syntax date_component : datetime
+
+private def parseDateTime : TSyntax `datetime -> MacroM (TSyntax `term)
+  | `(datetime| $date:date T $time:time) => do
+    `(Std.Time.PlainDateTime.mk $(← parseDate date) $(← parseTime time))
+  | `(datetime|$tm:date_component) => do
+    `(Std.Time.PlainDateTime.ofUTCTimestamp $(← parseComponent none none tm))
+  | syn => Macro.throwErrorAt syn "unsupported syntax"
+
+/--
+Category of offsets like `+00:00`.
+-/
+declare_syntax_cat offset
+
+/--
+Timezone in `+HH:mm`, or `-HH:mm` format.
+-/
+syntax ("+" <|> "-") date_component ":" date_component : offset
+
+private def parseOffset : TSyntax `offset -> MacroM (TSyntax `term)
+  | `(offset| +$hour:date_component:$minutes:date_component) => do
+    `(Std.Time.TimeZone.Offset.ofHoursAndMinutes $(← parseComponent (some 0) (some 24) hour) $(← parseComponent (some 0) (some 59) minutes))
+  | `(offset| -$hour:date_component:$minutes:date_component) => do
+    `(Std.Time.TimeZone.Offset.ofHoursAndMinutes (- $(← parseComponent (some 0) (some 24) hour)) (-$(← parseComponent (some 0) (some 59) minutes)))
+  | syn => Macro.throwErrorAt syn "unsupported syntax"
+
+/--
+Category of timezones like `Z`, `UTC`, `GMT` and `+HH:mm`.
+-/
+declare_syntax_cat zone
+
+/--
+Timezone in `+HHmm`, or `-HHmm` format.
+-/
+syntax (offset <|> "UTC" <|> "GMT" <|> "Z" <|> ("(" term ")")) : zone
```

Comment:
```suggestion
syntax (offset <|> &"UTC" <|> &"GMT" <|> &"Z" <|> ("(" term ")")) : zone
```

We probably do not want to make these date markers new keywords, so using no-reserved symbols (i.e., `&"..."`) here seems better.